src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScannerTest.java

```diff
@@ -536,4 +539,62 @@ public void testTripleParallel() throws Exception {
         Assert.assertTrue(pbs.unvisited.contains(exec.getNode("8")));
         Assert.assertTrue(pbs.unvisited.contains(exec.getNode("9")));
     }
+
+    private void testParallelFindsLast(WorkflowJob job, String semaphoreName) throws Exception {
+        ForkScanner scan = new ForkScanner();
+        ChunkFinder labelFinder = new LabelledChunkFinder();
+
+        System.out.println("Testing that semaphore step is always the last step for chunk with "+job.getName());
+        WorkflowRun run  = job.scheduleBuild2(0).getStartCondition().get();
+        SemaphoreStep.waitForStart(semaphoreName+"/1", run);
+
+            /*if (run.getExecution() == null) {
+                Thread.sleep(1000);
+            }*/
+
+        TestVisitor visitor = new TestVisitor();
+        scan.setup(run.getExecution().getCurrentHeads());
+        scan.visitSimpleChunks(visitor, labelFinder);
+        TestVisitor.CallEntry entry = visitor.calls.get(0);
+        Assert.assertEquals(TestVisitor.CallType.CHUNK_END, entry.type);
+        FlowNode lastNode = run.getExecution().getNode(Integer.toString(entry.ids[0]));
+        Assert.assertEquals("Wrong End Node: ("+lastNode.getId()+") "+lastNode.getDisplayName(), "semaphore", lastNode.getDisplayFunctionName());
+
+        SemaphoreStep.success(semaphoreName+"/1", null);
+        r.waitForCompletion(run);
+    }
+
+    @Issue("JENKINS-38536")
+    @Test
+    public void testParallelCorrectEndNodeForVisitor() throws Exception {
+        // Verify that SimpleBlockVisitor actually gets the *real* last node not just the last declared branch
+        WorkflowJob jobPauseFirst = r.jenkins.createProject(WorkflowJob.class, "PauseFirst");
+        jobPauseFirst.setDefinition(new CpsFlowDefinition("" +
+                "stage 'primero'\n" +
+                "parallel 'wait' : {sleep 1; semaphore 'wait1';}, \n" +
```

Comment:
Generally you should not need to use `sleep` steps if you are using `semaphore` correctly.